app/controllers/concerns/foreman/controller/smart_proxy_auth.rb

```diff
@@ -50,11 +50,29 @@ def require_smart_proxy_or_login(features = nil)
   def auth_smart_proxy(proxies = SmartProxy.all, require_cert = true)
     request_hosts = nil
     if request.ssl?
-      dn = request.env[Setting[:ssl_client_dn_env]]
-      if dn && dn =~ /CN=([^\s\/,]+)/i
+      # If we have the client certficate in the request environment we can extract the dn and sans from there
+      # if not we use the dn in the request environment
+      # SAN validation requires "SSLOptions +ExportCertData" in Apache httpd
+      if request.env.has_key?(Setting[:ssl_client_cert_env]) && request.env[Setting[:ssl_client_cert_env]].present?
+        logger.debug "Examining client certificate to extract dn and sans"
+        cert_raw = request.env[Setting[:ssl_client_cert_env]]
+        certificate = CertificateExtract.new(cert_raw)
+        logger.debug "Client sent certificate with subject '#{certificate.subject}' and subject alt names '#{certificate.subject_alternative_names.inspect}'"
+      else
+        dn = request.env[Setting[:ssl_client_dn_env]]
+      end
+
+      if (dn && dn =~ /CN=([^\s\/,]+)/i) || certificate
         verify = request.env[Setting[:ssl_client_verify_env]]
         if verify == 'SUCCESS'
-          request_hosts = [$1]
+          # If the client sent certificate contains a subject or sans, use them for request_hosts, else fall back to the dn set in the request environment
+          request_hosts = []
+          unless certificate
+            request_hosts << $1 if $1
+          else
+            request_hosts << certificate.subject if certificate.subject and !certificate.subject_alternative_names
```

Comment:
Use `&&` instead of `and`.

It might be better to rewrite this as a more basic if/else, since this line checks for absence of SANs, the next checks for presence.

<pre>
if certificate.subject_alternative_names
  request_hosts += certificate.subject_alternative_names
elsif certificate.subject
  request_hosts << certificate.subject
end
</pre>

or similar.